presto-rcfile/src/main/java/io/prestosql/rcfile/TimestampUtils.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.rcfile;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.type.LongTimestamp;
+import io.prestosql.spi.type.TimestampType;
+
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+
+import static io.prestosql.spi.type.Timestamps.MICROSECONDS_PER_SECOND;
+import static io.prestosql.spi.type.Timestamps.PICOSECONDS_PER_MICROSECOND;
+import static io.prestosql.spi.type.Timestamps.PICOSECONDS_PER_NANOSECOND;
+import static java.lang.Math.floorDiv;
+import static java.lang.Math.floorMod;
+import static java.lang.Math.toIntExact;
+
+public final class TimestampUtils
+{
+    private TimestampUtils() {}
+
+    public static LocalDateTime getLocalDateTime(TimestampType type, Block block, int position)
```

Comment:
I don't think this is a good abstraction. `LocalDateTime` seems fairly expensive in terms of object allocation and encoding (there's a reason we don't use it everywhere and continue to use Joda-Time). We're also doing the type switching for every value, rather than once for the block or writer instance.

Instead, for the binary encoding, we could inline this into the timestamp writing code (pulling the type switch outside of the block encoding loop). For text encoding, performance is less important, since it's already expensive.

It's possible that this doesn't actually have a performance impact, or that we don't care about RCFile anymore, but since we seem to be introducing a regression to the existing code for millis precision, I'd like to make sure this is an explicit decision (possibly based on data) rather than accidental. @dain probably has more thoughts here.